Input guardrail tripwire loses input_guardrail_results in non-streaming runs

### Please read this first

- **Have you read the docs?** Yes — [Guardrails](https://openai.github.io/openai-agents-python/guardrails/), [Results](https://openai.github.io/openai-agents-python/results/), and `.agents/references/runner-lifecycle.md`.
- **Have you searched for related issues?** Yes. Searched open/closed issues and PRs for `input_guardrail_results`, `InputGuardrailTripwireTriggered run_data`, `guardrail results run_data`, `RunErrorDetails input_guardrail_results`, `tripwire results missing`, `guardrail output_info lost`, and `input guardrail results streaming parity`. The closest prior work is #3982 and #4004 (cancelling in-flight tasks when an input guardrail trips) and #2688 (stopping streamed tool execution after a known tripwire) — all about task cancellation, none about the results reported on the raised exception.

### Affected component

`AgentRunner.run` / `Runner.run` / `Runner.run_sync` (non-streaming input-guardrail path) — `src/agents/run_internal/guardrails.py::run_input_guardrails` and its call sites in `src/agents/run.py`.

### Describe the bug

When an input guardrail tripwire fires, `InputGuardrailTripwireTriggered.run_data.input_guardrail_results` is **empty** for `Runner.run()` / `Runner.run_sync()`, but correctly contains every completed guardrail result for `Runner.run_streamed()`.

`run_input_guardrails()` accumulates results in a local list and raises `InputGuardrailTripwireTriggered(result)` as soon as a tripwire fires, discarding the accumulated list. `src/agents/run.py` only merges those results into the run-level `input_guardrail_results` on the success path (`input_guardrail_results.extend(sequential_results)`), so the `RunErrorDetails` attached to the exception is built from an empty list.

The streamed path uses `run_input_guardrails_with_queue()`, which assigns `streamed_result.input_guardrail_results` before propagating the tripwire, so streamed consumers see all results.

This means a non-streaming caller cannot read the `output_info` of the guardrails that ran — including the ones that passed — from the raised exception. `exc.guardrail_result` still exposes the single tripping result, so the loss is specifically the full, documented run-state view.

### Debug information

- Agents SDK version: `main` @ `c3f1781d56e8f1249a01674f18ba1f3e44a16dce` (latest release tag `v0.19.1`)
- Python version: 3.13.5
- OS: macOS 15.7.3 (Darwin 24.6.0)
- No OpenAI API key, no network access, and no paid model call are needed — the repro uses the repository's existing `tests/fake_model.py::FakeModel`.

### Repro steps

Save as `tests/test_repro_guardrail_results.py` and run `uv run pytest tests/test_repro_guardrail_results.py -q`:

```python
from __future__ import annotations

import asyncio
from typing import Any

import pytest

from agents import (
    Agent,
    GuardrailFunctionOutput,
    InputGuardrail,
    InputGuardrailTripwireTriggered,
    Runner,
)

from .fake_model import FakeModel
from .test_responses import get_text_message


def _guardrail(name: str, trip: bool, delay: float = 0.0) -> InputGuardrail[Any]:
    async def fn(ctx, agent, inp) -> GuardrailFunctionOutput:
        if delay:
            await asyncio.sleep(delay)
        return GuardrailFunctionOutput(output_info=name, tripwire_triggered=trip)

    return InputGuardrail(guardrail_function=fn, name=name)


def _agent(model: FakeModel) -> Agent[Any]:
    return Agent(
        name="A",
        model=model,
        input_guardrails=[
            _guardrail("passes", trip=False),
            _guardrail("trips", trip=True, delay=0.02),
        ],
    )


@pytest.mark.asyncio
async def test_streamed_exposes_input_guardrail_results() -> None:
    """Baseline: the streamed path already reports every guardrail result."""
    model = FakeModel()
    model.set_next_output([get_text_message("done")])
    result = Runner.run_streamed(_agent(model), input="hi")
    with pytest.raises(InputGuardrailTripwireTriggered) as exc_info:
        async for _ in result.stream_events():
            pass

    run_data = exc_info.value.run_data
    assert run_data is not None
    assert [r.guardrail.get_name() for r in run_data.input_guardrail_results] == ["passes", "trips"]


@pytest.mark.asyncio
async def test_non_streamed_exposes_input_guardrail_results() -> None:
    """Runner.run() must report the same guardrail results as run_streamed()."""
    model = FakeModel()
    model.set_next_output([get_text_message("done")])
    with pytest.raises(InputGuardrailTripwireTriggered) as exc_info:
        await Runner.run(_agent(model), input="hi")

    run_data = exc_info.value.run_data
    assert run_data is not None
    assert [r.guardrail.get_name() for r in run_data.input_guardrail_results] == ["passes", "trips"]
```

### Actual behavior

The streamed test passes; the non-streamed test fails deterministically (reproduced 3/3 runs):

```
FAILED tests/test_repro_guardrail_results.py::test_non_streamed_exposes_input_guardrail_results
E       AssertionError: assert [] == ['passes', 'trips']
E         Right contains 2 more items, first extra item: 'passes'
```

### Expected behavior

`InputGuardrailTripwireTriggered.run_data.input_guardrail_results` should list the guardrail results that completed before the run was aborted — including the tripping one — for `Runner.run()` and `Runner.run_sync()`, exactly as it already does for `Runner.run_streamed()`.

Two statements in the repository's own maintainer reference (`.agents/references/runner-lifecycle.md`) describe this expectation:

> Guardrail results are observable run state. Preserve them across handoffs, **error handlers**, streamed completion, and `RunState` round-trips.

> Streaming and non-streaming paths must produce equivalent final output, generated items, current agent, usage, **guardrail results**, session history, and interruption state for the same model behavior.

### Root-cause hypothesis (hypothesis, not verified as the maintainers' intent)

`run_input_guardrails()` in `src/agents/run_internal/guardrails.py` builds `guardrail_results` locally and raises before the caller can observe it. `src/agents/run.py` assigns the results into the run-level list only after a successful return (line ~1314, `input_guardrail_results.extend(sequential_results)`), so the `except BaseException` handler that populates `RunErrorDetails` sees an empty list.

### Proposed focused scope

Let `run_input_guardrails()` publish results into a caller-owned accumulator (or otherwise surface them on the tripwire path) so `src/agents/run.py` can attach them to `RunErrorDetails` before re-raising. No public API additions; `InputGuardrailTripwireTriggered.guardrail_result` and the exception type are unchanged. Regression tests would cover `Runner.run`, `Runner.run_sync`, and `Runner.run_streamed` for sequential and parallel guardrails.

### User impact

Non-streaming callers that inspect `err.run_data.input_guardrail_results` in an error handler — to log which checks ran, or to read `output_info` (scores, categories, redaction details) from guardrails that passed — always observe an empty list. Because `run_streamed()` behaves correctly, the same handler silently changes behavior depending on which `Runner` entry point is used.

I'd like to work on this and open a PR if that's welcome.

--- a/src/agents/run.py
+++ b/src/agents/run.py
@@ -797,16 +797,16 @@ class AgentRunner:
                         g for g in all_input_guardrails if not g.run_in_parallel
                     ]
                     parallel_guardrails = [g for g in all_input_guardrails if g.run_in_parallel]
-                    sequential_results: list[InputGuardrailResult] = []
                     if sandbox_runtime.enabled and sequential_guardrails:
                         # Blocking first-turn guardrails must run before sandbox prep so a tripwire
                         # can prevent session creation, startup, or live-session mutation.
                         try:
-                            sequential_results = await run_input_guardrails(
+                            await run_input_guardrails(
                                 starting_agent,
                                 sequential_guardrails,
                                 copy_input_items(original_input),
                                 context_wrapper,
+                                input_guardrail_results,
                             )
                         except InputGuardrailTripwireTriggered:
                             session_input_items_for_persistence = (
@@ -1221,11 +1221,12 @@ class AgentRunner:
                         if current_turn <= 1:
                             try:
                                 if sequential_guardrails:
-                                    sequential_results = await run_input_guardrails(
+                                    await run_input_guardrails(
                                         starting_agent,
                                         sequential_guardrails,
                                         copy_input_items(original_input),
                                         context_wrapper,
+                                        input_guardrail_results,
                                     )
                             except InputGuardrailTripwireTriggered:
                                 session_input_items_for_persistence = (
@@ -1240,7 +1241,6 @@ class AgentRunner:
                                 )
                                 raise
 
-                            parallel_results: list[InputGuardrailResult] = []
                             model_task = asyncio.create_task(
                                 run_single_turn(
                                     bindings=current_bindings,
@@ -1272,10 +1272,11 @@ class AgentRunner:
                                         parallel_guardrails,
                                         copy_input_items(original_input),
                                         context_wrapper,
+                                        input_guardrail_results,
                                     )
                                 )
                                 try:
-                                    parallel_results, turn_result = await asyncio.gather(
+                                    _, turn_result = await asyncio.gather(
                                         guardrail_task,
                                         model_task,
                                     )
@@ -1310,9 +1311,6 @@ class AgentRunner:
                                     raise
                             else:
                                 turn_result = await model_task
-
-                            input_guardrail_results.extend(sequential_results)
-                            input_guardrail_results.extend(parallel_results)
                         else:
                             turn_result = await run_single_turn(
                                 bindings=current_bindings,

--- a/src/agents/run_internal/guardrails.py
+++ b/src/agents/run_internal/guardrails.py
@@ -121,8 +121,14 @@ async def run_input_guardrails(
     guardrails: list[InputGuardrail[TContext]],
     input: str | list[TResponseInputItem],
     context: RunContextWrapper[TContext],
+    results_sink: list[InputGuardrailResult] | None = None,
 ) -> list[InputGuardrailResult]:
-    """Run input guardrails concurrently and raise on tripwires."""
+    """Run input guardrails concurrently and raise on tripwires.
+
+    Results are recorded into ``results_sink`` as each guardrail completes, including the
+    tripping result, so callers can report them even when this function raises. The streamed
+    path publishes the same results through `RunResultStreaming.input_guardrail_results`.
+    """
     if not guardrails:
         return []
 
@@ -133,10 +139,16 @@ async def run_input_guardrails(
 
     guardrail_results: list[InputGuardrailResult] = []
 
+    def record(result: InputGuardrailResult) -> None:
+        guardrail_results.append(result)
+        if results_sink is not None:
+            results_sink.append(result)
+
     try:
         for done in asyncio.as_completed(guardrail_tasks):
             result = await done
             if result.output.tripwire_triggered:
+                record(result)
                 for t in guardrail_tasks:
                     t.cancel()
                 await asyncio.gather(*guardrail_tasks, return_exceptions=True)
@@ -147,7 +159,7 @@ async def run_input_guardrails(
                     )
                 )
                 raise InputGuardrailTripwireTriggered(result)
-            guardrail_results.append(result)
+            record(result)
     except BaseException:
         # On any error (including a guardrail raising or the caller being cancelled),
         # cancel and await siblings so they don't leak past this function's return.

--- a/tests/test_guardrails.py
+++ b/tests/test_guardrails.py
@@ -1997,3 +1997,140 @@ async def test_output_guardrail_raise_cancels_siblings():
 
     assert sibling_cancelled.is_set(), "Sibling task should have been cancelled"
     assert not sibling_completed.is_set(), "Sibling task should not have completed"
+
+
+def _ordered_input_guardrails(
+    *, second_triggers: bool, second_raises: bool = False, run_in_parallel: bool = False
+) -> list[InputGuardrail[Any]]:
+    """Build two guardrails whose completion order is fixed by an explicit barrier."""
+    first_done = asyncio.Event()
+
+    async def first_fn(
+        context: RunContextWrapper[Any], agent: Agent[Any], input: str | list[TResponseInputItem]
+    ) -> GuardrailFunctionOutput:
+        first_done.set()
+        return GuardrailFunctionOutput(output_info="passes", tripwire_triggered=False)
+
+    async def second_fn(
+        context: RunContextWrapper[Any], agent: Agent[Any], input: str | list[TResponseInputItem]
+    ) -> GuardrailFunctionOutput:
+        await first_done.wait()
+        if second_raises:
+            raise RuntimeError("guardrail exploded")
+        return GuardrailFunctionOutput(output_info="second", tripwire_triggered=second_triggers)
+
+    return [
+        InputGuardrail(guardrail_function=first_fn, name="passes", run_in_parallel=run_in_parallel),
+        InputGuardrail(
+            guardrail_function=second_fn,
+            name="raises" if second_raises else "trips",
+            run_in_parallel=run_in_parallel,
+        ),
+    ]
+
+
+def _tripwire_agent(model: FakeModel, *, run_in_parallel: bool) -> Agent[Any]:
+    return Agent(
+        name="guardrail_results_agent",
+        model=model,
+        input_guardrails=_ordered_input_guardrails(
+            second_triggers=True, run_in_parallel=run_in_parallel
+        ),
+    )
+
+
+def _result_names(results: list[Any]) -> list[str]:
+    return [result.guardrail.get_name() for result in results]
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("run_in_parallel", [False, True])
+async def test_input_guardrail_tripwire_reports_results(run_in_parallel: bool):
+    """Runner.run() reports every completed guardrail result on the raised tripwire."""
+    model = FakeModel()
+    model.set_next_output([get_text_message("hello")])
+
+    with pytest.raises(InputGuardrailTripwireTriggered) as exc_info:
+        await Runner.run(_tripwire_agent(model, run_in_parallel=run_in_parallel), "test input")
+
+    run_data = exc_info.value.run_data
+    assert run_data is not None
+    assert _result_names(run_data.input_guardrail_results) == ["passes", "trips"]
+    assert exc_info.value.guardrail_result.guardrail.get_name() == "trips"
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("run_in_parallel", [False, True])
+async def test_input_guardrail_tripwire_reports_results_streamed(run_in_parallel: bool):
+    """The streamed path reports the same results, including on the streamed result object."""
+    model = FakeModel()
+    model.set_next_output([get_text_message("hello")])
+
+    result = Runner.run_streamed(
+        _tripwire_agent(model, run_in_parallel=run_in_parallel), "test input"
+    )
+    with pytest.raises(InputGuardrailTripwireTriggered) as exc_info:
+        async for _ in result.stream_events():
+            pass
+
+    run_data = exc_info.value.run_data
+    assert run_data is not None
+    assert _result_names(run_data.input_guardrail_results) == ["passes", "trips"]
+    assert _result_names(result.input_guardrail_results) == ["passes", "trips"]
+
+
+def test_input_guardrail_tripwire_reports_results_sync():
+    """Runner.run_sync() matches the async entry points."""
+    model = FakeModel()
+    model.set_next_output([get_text_message("hello")])
+
+    with pytest.raises(InputGuardrailTripwireTriggered) as exc_info:
+        Runner.run_sync(_tripwire_agent(model, run_in_parallel=False), "test input")
+
+    run_data = exc_info.value.run_data
+    assert run_data is not None
+    assert _result_names(run_data.input_guardrail_results) == ["passes", "trips"]
+
+
+@pytest.mark.asyncio
+async def test_input_guardrail_results_reported_on_success():
+    """Passing guardrails still land on the successful result exactly once."""
+    model = FakeModel()
+    model.set_next_output([get_text_message("hello")])
+    agent = Agent(
+        name="guardrail_results_agent",
+        model=model,
+        input_guardrails=[
+            InputGuardrail(
+                guardrail_function=get_sync_guardrail(triggers=False),
+                name="blocking",
+                run_in_parallel=False,
+            ),
+            InputGuardrail(
+                guardrail_function=get_sync_guardrail(triggers=False),
+                name="parallel",
+                run_in_parallel=True,
+            ),
+        ],
+    )
+
+    result = await Runner.run(agent, "test input")
+
+    assert _result_names(result.input_guardrail_results) == ["blocking", "parallel"]
+
+
+@pytest.mark.asyncio
+async def test_input_guardrail_exception_reports_completed_results():
+    """A guardrail raising a non-tripwire error still preserves earlier results."""
+
+    collected: list[Any] = []
+    with pytest.raises(RuntimeError, match="guardrail exploded"):
+        await run_input_guardrails(
+            Agent(name="t"),
+            _ordered_input_guardrails(second_triggers=False, second_raises=True),
+            "test input",
+            RunContextWrapper(context=None),
+            collected,
+        )
+
+    assert _result_names(collected) == ["passes"]

